zrml/swaps/src/lib.rs

```diff
@@ -1850,17 +1855,22 @@ mod pallet {
         }
 
         fn open_pool(pool_id: PoolId) -> Result<Weight, DispatchError> {
-            Self::mutate_pool(pool_id, |pool| {
-                ensure!(
-                    pool.pool_status == PoolStatus::Initialized,
-                    Error::<T>::InvalidStateTransition
-                );
-                pool.pool_status = PoolStatus::Active;
-                Ok(())
-            })?;
+            let asset_len =
+                <Pools<T>>::try_mutate(pool_id, |pool| -> Result<u32, DispatchError> {
+                    let pool = if let Some(el) = pool {
+                        el
+                    } else {
+                        return Err(Error::<T>::PoolDoesNotExist.into());
+                    };
+                    ensure!(
+                        pool.pool_status == PoolStatus::Initialized,
+                        Error::<T>::InvalidStateTransition
+                    );
+                    pool.pool_status = PoolStatus::Active;
+                    Ok(pool.assets.len() as u32)
+                })?;
```

Comment:
[nitpick] Too bad that you have to copy the `mutate_pool` code, but I don't see any other way. Suppose you did this...

```rust
Self::mutate_pool(...);
let pool = Pools::<T>::get(pool_id);
Ok(pool.assets.len() as u32)
```

Would that result in an additional read after substrate does its magic optimization?

Reply:
34aa62a6765e7eafc8e93cefecba72f7a28dfb28